app/services/create_work_package_service.rb

```diff
@@ -0,0 +1,52 @@
+#-- encoding: UTF-8
+#-- copyright
+# OpenProject is a project management system.
+# Copyright (C) 2012-2015 the OpenProject Foundation (OPF)
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See doc/COPYRIGHT.rdoc for more details.
+#++
+
+class CreateWorkPackageService
+  attr_accessor :user, :project, :work_package
+
+  def initialize(user:, project:, send_notifications: true)
+    self.user = user
+    self.project = project
+
+    WorkPackageObserver.instance.send_notification = send_notifications
+  end
+
+  def create
+    hash = {
+      project: project,
+      author: user,
+      type: project.types.where(is_default: true).first || project.types.first
+    }
+    self.work_package = project.add_work_package(hash)
+  end
+
+  def save
+    work_package.save
```

Comment:
@Azure7111, @NobodysNightmare I am still struggling with actually using services myself so this is more of an opinion than an advice.

I think initialization of the service should take the context, almost what @Azure7111 did. To me, the context is the user and the project. That will enable reuse of the service for creating multiple work packages.

On top of that, if we go with very specific services (CreateWorkPackageService as opposed to WorkPackageService) which I like so far, there should probably be only one other instance method: `call`. The method name is chosen because it is the same as the one Procs use in ruby. Thus, the implementation becomes easily interchangeable. That method should take the concrete information the service need to perform the task. In this case, it would be a dictionary (attribute, value). I'd put the notification in there as well. The work package should not be exposed to the outside but it automatically no longer is if the work package is taken into the call method and the Service does not keep it. 

IMO, the work package should be completely saved and the action completed once the service is called. 

That would mean, that instead of writing this in the controller:
```
   work_package = create_service.create

   write_work_package_attributes work_package

   if write_request_valid?(work_package, WorkPackages::CreateContract) &&
       create_service.save
  ....
```
you'd write:
```
attributes_from_request = Representer.parse(request)

if create_service.call(attributes_from_request)
 ...
````

That would mean that the contract is known to the Service and not in the controller.

Additionally, I think that the call to the WorkPackageObserver should be encapsulated within another Service that is injected into the WorkPackageCreateService.

Now, having written all this I am quite aware that it will be quite hard to refactor what we have in the controller to work in the service as well. As to what can be done now, we better talk in person.